croquemort/tools.py

```diff
@@ -75,17 +51,66 @@ def extract_filters(querystring_dict):
     return filters, excludes
 
 
-def apply_filters(results, filters, excludes):
-    """Return filtered results."""
-    if filters:
-        if all(results.get(prop) == value
-               for prop, value in filters.items()):
-            return results
+def apply_filters(data, filters, excludes):
+    """Return filtered data."""
+    filters = filters.copy()
+    excludes = excludes.copy()
+    has_domain_filter = 'domain' in filters
+    has_domain_exclude = 'domain' in excludes
+    has_domain = has_domain_filter or has_domain_exclude
+    filtered_domain = (
+        has_domain_filter
+        and urlparse(data['url']).netloc == filters.pop('domain'))
+    excluded_domain = (
+        has_domain_exclude
+        and urlparse(data['url']).netloc == excludes.pop('domain'))
+    kept_domain = ((has_domain_filter and filtered_domain)
+                   or (has_domain_exclude and not excluded_domain))
+    # Only filter the domain and quick return if no other filters.
+    if has_domain and not (filters or excludes):
+        if kept_domain:
+            return data
+        else:
+            return
+
+    has_props = all(data.get(prop) == value
+                    for prop, value in filters.items())
+    has_not_props = all(data.get(prop) != value
+                        and data.get(prop) is not None
+                        for prop, value in excludes.items()
+                        if prop in data)
+    if filters and excludes:
```

Comment:
Last thought, ``has_domain and not kept_domain`` seems to be factorizable:
```python
if has_domain and not kept_domain:
    return
elif filters and excludes and not has_props or not has_not_props:
    return
elif filters and not has_props:
    return
elif excludes and not has_not_props:
    return

return data
```
or If the case without both `filters` and `excludes` is acceptable
```python
if any(filters, excludes) and has_domain and not kept_domain:
    return
elif filters and excludes and not has_props or not has_not_props: # May needs parenthesis
    return
elif filters and not has_props:
    return
elif excludes and not has_not_props:
    return

return data
```